### ISSUE TITLE
new check: LeadingAsteriskAlignCheck

### ISSUE THREAD

**Issue by pbludov** [checkstyle/checkstyle#6723]
Checks that the leading asterisk on each line is vertically aligned.

It checks for vertical alignment of the leading asterisk, it ensures that the leading asterisk is vertically aligned properly under the first `*` of the starting javadoc block ( /** ). The Check also ensures the alignment of the ending block of the javadoc ( */ ). The ending block's `*` should be aligned vertically under the first `*` of `/**`.

Example:

```java
/**
 * properly aligned asterisk
 */
```

Violation example:

```java
/**
* // violation
  *  // violation
        * // violation
*/ // violation
``` 

If there is no leading asterisk for a line then that line will be ignored. If there is text before the `*/` then alignment for it will also be ignored. There is already a Check which checks for [missing leading asterisk](https://checkstyle.sourceforge.io/checks/javadoc/javadocmissingleadingasterisk.html) in javadoc. You have also suggested to ignore the line which does not contain leading asterisk, [here](https://github.com/checkstyle/checkstyle/issues/6723#issuecomment-493951547):

> We would just have to look for the first * starting the line and ignore the line if it's first character is not * and is a non-whitespace character.

Example:

```java
/**
          There is no leading asterisk,
So this will not cause error */
```

If there is a leading asterisk at last line i.e. the line containing the ending block ( */ ) then the leading asterisk's alignment will be considered.

Example:

```java
/**
 *
 * the leading asterisk is aligned properly, no error */

/**
 *
     * the leading asterisk is not aligned properly, will cause an error */
```

========

**Old Ideas:**

* property `align`: `left` for r-shaped comments, `right` for 1-shaped comments

*  align=left (default)
```
/**
 * OK (asterisk's column == left asterisk's column from `/**` )
   Missing asterisk ignored
 ** violation
  * violation
* violation
     * violation
 */
```
*  align=right
```
/**
  * OK (asterisk's column == right asterisk's column from `/**` )
   Missing asterisk ignored
 ** violation
 * violation
* violation
     * violation
 */

**Comment by rnveach** [comment]
@pbludov Please update first post and show some violation cases with each option.
Does check care about double `*`s if it covers both columns? I assume violations will be shown if `*` is more then 1 over and less then 1.
Will this check provide no violations if asterisk is missing? We shouldn't duplicate https://github.com/checkstyle/checkstyle/issues/6722 .

**Comment by pbludov** [comment]
> @pbludov Please update first post and show some violation cases with each option.

done
> Does check care about double `*`s if it covers both columns? I assume violations will be shown if `*` is more then 1+ after or less then 1+ behind.

Yes. There should be exactly one `*` in the right position.
> Will this check provide no violations if asterisk is missing? We shouldn't duplicate #6722 .

No. It should iterate over `LEADING_ASTERISK` token and check the column.

**Comment by rnveach** [comment]
@romani Please review.

**Comment by romani** [comment]
> align: left for r-shaped comments, right for 1-shaped comments

asterisk could be under "/" and will looks ok, I am not sure who might use it, but  it still ok as whole comment block is properly indented (from general indentation view). 
```
/**
* OK
*/
```
should we do:
align: left for align under '/' , center - align under '*', right - align under second '*' ([javadoc identificator](https://checkstyle.org/writingjavadocchecks.html#What_is_Javadoc_comment))

**Comment by pbludov** [comment]
> asterisk could be under "/" and will looks ok, I am not sure who might use it, but it still ok as whole comment block is properly indented (from general indentation view).

I've never seen that style. We can start with `center` and `right`, then add `left` if anyone asks for it

> attention: looks like we should be clear in aligning of `*/`, so we target only leading asterisk or ....?

For now, this check can't handle `BLOCK_COMMENT_END`, as it is not a javadoc token. I think we need another check for this token.

**Comment by romani** [comment]
>then add left if anyone asks for it

Yes

>can't handle BLOCK_COMMENT_END, as it is not a javadoc token. 

It's make clear in documentation, as majority of user do not know such javadoc details.

@rnveach, what are your thoughts on this ?

**Comment by rnveach** [comment]
> can't handle BLOCK_COMMENT_END, as it is not a javadoc token.

This brings up a good point.
Why does this need to be a javadoc check? Block comments also have similar issues with `*` positioning issues if it spans multiple lines. It is not just isolated to the javadoc.
Why don't we just make this validate all block comments instead of only javadocs? Yes, Javadocs are where most people use block comments the most but I don't see much difference in extending this to all block comments besides we won't be able to use the Javadoc parser. We would just have to look for the first `*` starting the line and ignore the line if it's first character is not `*` and is a non-whitespace character.

**Comment by romani** [comment]
I am ok to have it applicable as general block comment Check. We do not need javadoc parser to make validation.

**Comment by rnveach** [comment]
@romani If we are all in agreement, lets approve the issue.

This applies to the sun checks.
https://checkstyle.org/styleguides/sun-code-conventions-19990420/CodeConventions.doc4.html#a16838
> subsequent doc comment lines each have 1 space of indentation (to vertically align the asterisks).

Didn't review google.

**Comment by bossever** [comment]
I would like to work on this. 
Before I start, I would like to make sure of a few things - 

### 1.
As it will **NOT** be a javadoc check, does that mean the properties left and right will behave like this?
```
/* Block comment begin
* align = left
 * align = right
*/

/** Javadoc begin
* align = left
 * align = right
*/
```
   Or will they behave differently for block comments and javadoc?
_____________________________
### 2.
This check will **NOT** raise violations for any line that has missed the leading asterisk, regardless of indentation/whitespace. (whether that is in a block comment or a javadoc)
_____________________________
### 3.
This check will **NOT** check the comment end token.

**Comment by pbludov** [comment]
> I would like to work on this.
> Before I start, I would like to make sure of a few things -

1. To begin with, let behavior be the same. I don't see any reason for the differences here.
2. Yes. No leading asterisk == no violations. We have a dedicated check for this.
3. If this is not a Javadoc check, then it is possible check the alignment of `COMMENT_END_TOKEN`.

**Comment by bossever** [comment]
Since this check will be dealing with various methods of indentation (tabs vs spaces), I have added an option to specifiy the tab indentation level - `tabIndent` which will have a default value of 4.
Leave a thumbs up if this is the correct way to deal with this.

**Comment by Zopsss** [comment]
Hey @romani @rnveach @pbludov I was thinking about working on this issue. I have cherry picked the changes to my local setup. I've read the implementation code of this check and I guess the functionality of the check is implemented as discussed here, it is just giving some errors on `mvn clean verify` which I'm trying to solve. There is some documentation work left that's needs to be done, is there something else too to work on?

And also the `tabSize` property introduced here https://github.com/checkstyle/checkstyle/issues/6723#issuecomment-831916856. Are we going to implement it or not? I'm not able to understand what does it do? How it will be used with the Check? Can anyone please explain me what will be the usecase of this property?

**Comment by nrmancuso** [comment]
We should create an initial version of this check that only supports the style that we use, and uses the Javadoc AST. Block comments have no need for such checks, and if they do, it should be in their own check that uses the Java AST, with comments.

Consider:
 - it will be much easier to support only one style at first to get check into actual usage
 - block comments can appear embedded in any java code, support for this will be more difficult
 - check implementation will be much easier and reliable with AST

**Comment by romani** [comment]
I am ok with step by step improvement in implementation/design.

**Comment by Zopsss** [comment]
> We should create an initial version of this check that only supports the style that we use, and use the Javadoc AST. 

I'm also okay with this. But before transitioning to the Javadoc Check, I want to make some things clear:

1. We don't need to check that if the Javadoc line contains leading asterisk or not as it is already covered by [JavadocMissingLeadingAsterisk](https://checkstyle.sourceforge.io/checks/javadoc/javadocmissingleadingasterisk.html#JavadocMissingLeadingAsterisk)
2. It is not possible to check if the line contains consecutive leading asterisks with `DetailNode` so we will check that if the first leading asterisk is properly aligned or not. We will not throw an error for having consecutive leading asterisks.
3. It is also not possible to check the align of leading asterisk for end block of Javadoc comment with `DetailNode`
4. Now that we're moving from checking block comments & javadoc comments to checking only javadoc comment, should we consider this approach for alignment: https://github.com/checkstyle/checkstyle/issues/6723#issuecomment-493777285 ? Or we should ignore the `left` alignment as mentioned here: https://github.com/checkstyle/checkstyle/issues/6723#issuecomment-493790846 ?


And I just wanted to let you know that my mid-term exams are coming soon so the new implementation might take some time but I'll try to complete it as soon as possible. @romani @nrmancuso

**Comment by nrmancuso** [comment]
@Zopsss thanks a lot for your summary here. We would need 100% alignment from maintainers before we proceed with a different approach. @rnveach are you good with my proposal above?

As for items from https://github.com/checkstyle/checkstyle/issues/6723#issuecomment-2001955545:

>We don't need to check that if the Javadoc line contains leading asterisk or not as it is already covered by [JavadocMissingLeadingAsterisk](https://checkstyle.sourceforge.io/checks/javadoc/javadocmissingleadingasterisk.html#JavadocMissingLeadingAsterisk)

Yes, this is good. For the most part, checks should compliment each other like this, and not overlap (which can cause conflicts). 

>Now that we're moving from checking block comments & javadoc comments to checking only javadoc comment, should we consider this approach for alignment: https://github.com/checkstyle/checkstyle/issues/6723#issuecomment-493777285 ? Or we should ignore the left alignment as mentioned here: https://github.com/checkstyle/checkstyle/issues/6723#issuecomment-493790846 ?

From https://github.com/checkstyle/checkstyle/issues/6723#issuecomment-1995177320:

>We should create an initial version of this check that only supports the style that we use

This means no options for alignment at all; I think we would consider what is currently used in our project to be "right" alignment for stylistic purposes.

**Comment by nrmancuso** [comment]
>So does that mean we no longer need the option property?

For initial implementation, no. We should check for whatever the style we currently use in Checkstyle is - we should dogfood new checks ourselves, and get actual user feedback to drive extension of the check.

**Comment by rnveach** [comment]
> @rnveach are you good with my proposal above?

I am considering this to mean https://github.com/checkstyle/checkstyle/issues/6723#issuecomment-1995177320 .

I agree except with items mentioned below.

> Block comments have no need for such checks, and if they do, it should be in their own check that uses the Java AST, with comments

I disagree that block comments don't need this, but I agree they should be in their own AST check.

I found it better to use block comments over single line comments for sentences, especially those over multiple lines. Auto formatters will correct the line width, if it is ever updated, for you when you expand or shorten existing lines and move the text around for you. It made it easier to read, and you wouldn't have oddly shapped lines with some much shorter than others.

Even with block comments the `*` on subsequent lines could not line up with the `*` in the `/*` or the `/` and still be crazly all over the place.

> Support for sun, google checks, etc. should be done in the scope of another issue, once check is proven.

While I agree with this, I think we should be aware of the needs of the other issue so we aren't creating a check that cannot be modified to solve the other issue.

**Comment by Zopsss** [comment]
> Even with block comments the `*` on subsequent lines could not line up with the `*` in the `/*` or the `/` and still be crazly all over the place. If it isn't enforce by some check, people will do whatever they want.

Yes, we can create a separate check for this, similar to this check but for block comments.
> While I agree with this, I think we should be aware of the needs of the other issue so we aren't creating a check that cannot be modified to solve the other issue.

Don't worry, I will try to keep the check as simple as possible and make sure it could be extensible to implement new functionalities in the future.

**Comment by nrmancuso** [comment]
I am still not clear on sun/google requirements, we should open the issue with all details on this before we proceed here.

**Comment by Zopsss** [comment]
#### Google Style Guide:

We need to check the block comment's leading asterisk alignment as mentioned in `4.8.6.1 Block comment style` ( we can create a new check for this ) :-
https://google.github.io/styleguide/javaguide.html#s4.8.6-comments

We also need to check the Javadoc comment's leading asterisk alignment as mentioned here ( part of our check ) :-
https://google.github.io/styleguide/javaguide.html#s7.1.1-javadoc-multi-line

---

#### Sun Style Guide:

We need to check Javadoc comment's leading asterisk as mentioned here:
https://checkstyle.org/styleguides/sun-code-conventions-19990420/CodeConventions.doc4.html#a16838
>The first line of doc comment (/**) for classes and interfaces is not indented; subsequent doc comment lines each have 1 space of indentation (to vertically align the asterisks)



We also need to check Javadoc's indentation as mentioned here ( we can add this functionality once the initial implementation of this check is done, we can create a new issue for adding this functionality ) :-
> Members, including constructors, have 4 spaces for the first doc comment line and 5 spaces thereafter.

The `5.1.1 Block Comments` section talks about Block Comments, but there is no rule defined for leading asterisk's alignment, though there is something mentioned about indentation :- 
> Block comments can start with /*-, which is recognized by indent(1) as the beginning of a block comment that should not be reformatted. Example:

But I didn't understand this statement that much so I'm not so sure about this.

**Comment by Zopsss** [comment]
> we should open the issue with all details on this before we proceed here.

I guess we should first finish the initial implementation then we should open issues on google & sun style guides. At that time we will also have the implemented check so the new functionalities will be implemented much faster.

**Comment by Zopsss** [comment]
@nrmancuso

**Comment by nrmancuso** [comment]
> @nrmancuso

Yes?

**Comment by Zopsss** [comment]
> My point from [#6723 (comment)](https://github.com/checkstyle/checkstyle/issues/6723#issuecomment-2004734460) still stands, we absolutely must track all work by issue, and this issue will not be able to be closed without transferring related work to a new issue, period.

Yes we will also add support for sun/google style guide once the basic implementation of this check is done. The basic implementation follows the rules mentioned in both style guides, which is that we need to vertically align the leading asterisk under the first asterisk of the Javadoc comment, I mean in the center.

> Your comment at [#6723 (comment)](https://github.com/checkstyle/checkstyle/issues/6723#issuecomment-2004898322) is nearly the entire issue description, and I don’t want it to get lost in here without an issue to track it.

I will create a new issue for this soon, I guess we don't need to modify our check as the initial implementation covers the rules mentioned in the style guide but we'll discuss all this stuff on the new issue. Till then I'll try to finish working on the implementation of the Check. @nrmancuso

**Comment by romani** [comment]
fix is merged
